Bottom: 30551361df1d9dc94d177073ae1c40a17c4767ad
Top:    8afeba49e1d9250b08e7dbcc08443a8b1ca55079
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-02-15 15:32:29 +0800

virito-net: keep old xdp_queue_paris during reset


---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 11e2853..f9a5431 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -1775,7 +1775,7 @@ static int virtnet_xdp_set(struct net_device *dev, struct bpf_prog *prog)
 	unsigned long int max_sz = PAGE_SIZE - sizeof(struct padded_vnet_hdr);
 	struct virtnet_info *vi = netdev_priv(dev);
 	struct bpf_prog *old_prog;
-	u16 oxdp_qp, xdp_qp = 0, curr_qp;
+	u16 xdp_qp = 0, curr_qp;
 	int i, err;
 
 	if (virtio_has_feature(vi->vdev, VIRTIO_NET_F_GUEST_TSO4) ||
@@ -1819,18 +1819,16 @@ static int virtnet_xdp_set(struct net_device *dev, struct bpf_prog *prog)
 		goto virtio_queue_err;
 	}
 
-	oxdp_qp = vi->xdp_queue_pairs;
-
 	/* Changing the headroom in buffers is a disruptive operation because
 	 * existing buffers must be flushed and reallocated. This will happen
 	 * when a xdp program is initially added or xdp is disabled by removing
 	 * the xdp program resulting in number of XDP queues changing.
 	 */
 	if (vi->xdp_queue_pairs != xdp_qp) {
-		vi->xdp_queue_pairs = xdp_qp;
 		err = virtnet_reset(vi);
 		if (err)
 			goto virtio_reset_err;
+		vi->xdp_queue_pairs = xdp_qp;
 	}
 
 	netif_set_real_num_rx_queues(dev, curr_qp + xdp_qp);
@@ -1850,7 +1848,6 @@ static int virtnet_xdp_set(struct net_device *dev, struct bpf_prog *prog)
 	 * us so not much we can do here.
 	 */
 	dev_warn(&dev->dev, "XDP reset failure and queues unstable\n");
-	vi->xdp_queue_pairs = oxdp_qp;
 virtio_queue_err:
 	/* On queue set error we can unwind bpf ref count and user space can
 	 * retry this is most likely an allocation failure.
